src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -33368,635 +33971,3631 @@ .maxstack  3
+
+            string sourceB2 = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        object x = 1;
+                        object[] y = [2, 3];
+                        MyCollection<object> z = [x, ..y];
+                    }
+                }
+                """;
+            var comp = CreateCompilation([sourceB2, sourceA]);
+            comp.VerifyEmitDiagnostics(
+                // (7,35): error CS1950: The best overloaded Add method 'MyCollection<object>.Add(string)' for the collection initializer has some invalid arguments
+                //         MyCollection<object> z = [x, ..y];
+                Diagnostic(ErrorCode.ERR_BadArgTypesForCollectionAdd, "x").WithArguments("MyCollection<object>.Add(string)").WithLocation(7, 35),
+                // (7,35): error CS1503: Argument 1: cannot convert from 'object' to 'string'
+                //         MyCollection<object> z = [x, ..y];
+                Diagnostic(ErrorCode.ERR_BadArgType, "x").WithArguments("1", "object", "string").WithLocation(7, 35),
+                // (7,38): error CS1503: Argument 1: cannot convert from 'object' to 'string'
+                //         MyCollection<object> z = [x, ..y];
+                Diagnostic(ErrorCode.ERR_BadArgType, "..y").WithArguments("1", "object", "string").WithLocation(7, 38),
+                // (7,40): error CS1950: The best overloaded Add method 'MyCollection<object>.Add(string)' for the collection initializer has some invalid arguments
+                //         MyCollection<object> z = [x, ..y];
+                Diagnostic(ErrorCode.ERR_BadArgTypesForCollectionAdd, "y").WithArguments("MyCollection<object>.Add(string)").WithLocation(7, 40));
+        }
+
+        [Fact]
+        public void AddMethod_09()
+        {
+            // public struct MyCollection<T> : IEnumerable
+            // {
+            //     IEnumerator IEnumerable.GetEnumerator() => null;
+            //     public void Add(T x, params T y) => throw null;
+            // }
+            string sourceA = $$"""
+                .class public sealed MyCollection`1<T>
+                  extends [mscorlib]System.ValueType
+                  implements [mscorlib]System.Collections.IEnumerable
+                {
+                  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+                  .method private instance class [mscorlib]System.Collections.IEnumerator GetEnumerator()
+                  {
+                    .override [mscorlib]System.Collections.IEnumerable::GetEnumerator
+                    ldnull
+                    ret
+                  }
+                  .method public instance void Add(!T x, !T y)
+                  {
+                    .param [2]
+                    .custom instance void [mscorlib]System.ParamArrayAttribute::.ctor() = ( 01 00 00 00 ) 
```

Comment:
Should we have a version with `ParamCollectionAttribute`?

Reply:
> Should we have a version with `ParamCollectionAttribute`?

`AddMethod_ParamCollectionAttribute_03`